main.tf

```diff
@@ -0,0 +1,78 @@
+module "label" {
+  source     = "git::https://github.com/cloudposse/terraform-null-label.git?ref=tags/0.3.3"
+  namespace  = "${var.namespace}"
+  stage      = "${var.stage}"
+  name       = "${var.name}"
+  attributes = "${var.attributes}"
+  delimiter  = "${var.delimiter}"
+  tags       = "${var.tags}"
+}
+
+locals {
+  enabled                  = "${var.enabled == "true" ? true : false}"
+  remapped_label_id        = "${replace(module.label.id, var.delimiter, "_")}"
+  rsa                      = "${var.ssh_key_algorithm == "RSA" ? true : false}"
+  ecdsa                    = "${var.ssh_key_algorithm == "ECDSA" ? true : false}"
+  default_public_key_name  = "${local.remapped_label_id}_public_key"
+  default_private_key_name = "${local.remapped_label_id}_private_key"
+  public_key_name          = "${length(var.ssh_public_key_name) > 0 ? var.ssh_public_key_name : local.default_public_key_name}"
+  private_key_name         = "${length(var.ssh_private_key_name) > 0 ? var.ssh_private_key_name : local.default_private_key_name}"
+  ssh_public_key_ssm_path  = "${format("/%s/%s", var.ssm_path_prefix, local.public_key_name)}"
```

Comment:
`/` is not supported by `chamber`. We can write it from TF to SSM, but would not be able to read it using `chamber`. Since the module description says "to read it using chamber", it will not work.

```
Error: Failed to validate key: Failed to validate key name '/example/demo/organization_account_access_role'.  Only alphanumeric, dashes, fullstops and underscores are allowed for key names
```

Suggest adding `var.path_delimiter` and use it in all paths. It could be set to `/` if needed in some cases.

```
ssh_public_key_ssm_path  = "${format("%s%s%s%s", var.path_delimiter, var.ssm_path_prefix, var.path_delimiter, local.public_key_name)}"
```

Reply:
@aknysh Am not sure I follow here. Example:

```
module.label.null_resource.default: Creating...
  triggers.%:          "" => "5"
  triggers.attributes: "" => ""
  triggers.id:         "" => "cpco-testing-foo"
  triggers.name:       "" => "foo"
  triggers.namespace:  "" => "cpco"
  triggers.stage:      "" => "testing"
module.label.null_resource.default: Creation complete after 0s (ID: 6246099253291691392)
tls_private_key.default_rsa: Creating...
  algorithm:                  "" => "RSA"
  ecdsa_curve:                "" => "P224"
  private_key_pem:            "" => "<computed>"
  public_key_fingerprint_md5: "" => "<computed>"
  public_key_openssh:         "" => "<computed>"
  public_key_pem:             "" => "<computed>"
  rsa_bits:                   "" => "4096"
tls_private_key.default_rsa: Creation complete after 6s (ID: 4975b9856ebe3239799e1666f241e641784a8597)
aws_ssm_parameter.private_rsa_key: Creating...
  arn:            "" => "<computed>"
  description:    "" => "TLS Private Key"
  key_id:         "" => "<computed>"
  name:           "" => "/ssh_keys/cpco_testing_foo_private_key"
  overwrite:      "" => "true"
  tags.%:         "" => "3"
  tags.Name:      "" => "cpco-testing-foo"
  tags.Namespace: "" => "cpco"
  tags.Stage:     "" => "testing"
  type:           "" => "SecureString"
  value:          "<sensitive>" => "<sensitive>"
aws_ssm_parameter.public_rsa_key: Creating...
  arn:            "" => "<computed>"
  description:    "" => "TLS Public Key (OpenSSH - RSA)"
  key_id:         "" => "<computed>"
  name:           "" => "/ssh_keys/cpco_testing_foo_public_key"
  overwrite:      "" => "true"
  tags.%:         "" => "3"
  tags.Name:      "" => "cpco-testing-foo"
  tags.Namespace: "" => "cpco"
  tags.Stage:     "" => "testing"
  type:           "" => "String"
  value:          "<sensitive>" => "<sensitive>"
aws_ssm_parameter.public_rsa_key: Creation complete after 5s (ID: /ssh_keys/cpco_testing_foo_public_key)
aws_ssm_parameter.private_rsa_key: Creation complete after 5s (ID: /ssh_keys/cpco_testing_foo_private_key)

Apply complete! Resources: 4 added, 0 changed, 0 destroyed.

Outputs:

public_key = ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDQImg+H0Fuvk0hgLw3Pk8WBzPdjjjlIbmFwPptl9aTqrQDpn20QcU3CEcGp+MLeDhSmogYqBfpMKX/4gJQkacwHlxollEmbJDJLAbF/0LymARFdA/9uQThlVnTjlEMDMEhJGVv1fCkg45qEcSmUDz/zAgmUo81RWvFDBVfhQ/+HexI2UgSDqL29VRIWImK3WHUTzmABEir1JrdD0v8ltMAVdnhkW+eQdN2116NBWLl1Fdxz0augUHVz2LmB3UZWa+X3ntwa/uPeh+OcBothmIjoXM162p1q6amhari9cvgU+CeZTkibyooD4T9pHrWMNVVWcK7pSZl6iLVxD5HL+5XK7Lhtsk5/j3BsboFTm8gufrnBUQDTnQ5arLWQxBOweOTpxfgNHS5uB4mf920yzEI+A37H5kXZ6V7agowsGYes885Sw3BEgy0YCaeEsPs1fPPBew9VgryFMsQbj2tXjrkBGVw1ZLKiq3UwSVb6nop/jUpGyVHdqZ4cd6M8e8uucnHi4oO166vpdSPIMgio3auaL09W2TgyiGBDRdyecPYcqN6VZM88PQdul26b94eIgcKshbbRNAelHeXQe0aWkmXWvSLxSOexbOYKJt4k1S54WAKEJs7coFdsSTiPGFYgYelaLTZNB7Fsh+FYfBFDpR0ctZ4Cut2vG+7VXhtNdz0Gw==
```

```
 ✓   (cpco-testing-admin) terraform-aws-ssm-tls-ssh-key-pair ⨠  chamber list ssh_keys
Key				Version		LastModified	User
cpco_testing_foo_private_key	0		01-10 15:44:02	arn:aws:sts::126450723953:assumed-role/OrganizationAccountAccessRole/1547134955786336700
cpco_testing_foo_public_key	0		01-10 15:44:02	arn:aws:sts::126450723953:assumed-role/OrganizationAccountAccessRole/1547134955786336700
```

```
 ✓   (cpco-testing-admin) terraform-aws-ssm-tls-ssh-key-pair ⨠  chamber read ssh_keys cpco_testing_foo_public_key
Key				Value	Version		LastModified	User
cpco_testing_foo_public_key	ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDQImg+H0Fuvk0hgLw3Pk8WBzPdjjjlIbmFwPptl9aTqrQDpn20QcU3CEcGp+MLeDhSmogYqBfpMKX/4gJQkacwHlxollEmbJDJLAbF/0LymARFdA/9uQThlVnTjlEMDMEhJGVv1fCkg45qEcSmUDz/zAgmUo81RWvFDBVfhQ/+HexI2UgSDqL29VRIWImK3WHUTzmABEir1JrdD0v8ltMAVdnhkW+eQdN2116NBWLl1Fdxz0augUHVz2LmB3UZWa+X3ntwa/uPeh+OcBothmIjoXM162p1q6amhari9cvgU+CeZTkibyooD4T9pHrWMNVVWcK7pSZl6iLVxD5HL+5XK7Lhtsk5/j3BsboFTm8gufrnBUQDTnQ5arLWQxBOweOTpxfgNHS5uB4mf920yzEI+A37H5kXZ6V7agowsGYes885Sw3BEgy0YCaeEsPs1fPPBew9VgryFMsQbj2tXjrkBGVw1ZLKiq3UwSVb6nop/jUpGyVHdqZ4cd6M8e8uucnHi4oO166vpdSPIMgio3auaL09W2TgyiGBDRdyecPYcqN6VZM88PQdul26b94eIgcKshbbRNAelHeXQe0aWkmXWvSLxSOexbOYKJt4k1S54WAKEJs7coFdsSTiPGFYgYelaLTZNB7Fsh+FYfBFDpR0ctZ4Cut2vG+7VXhtNdz0Gw==
				0	01-10 15:44:02	arn:aws:sts::126450723953:assumed-role/OrganizationAccountAccessRole/1547134955786336700
```

Reply:
ah, ok then.
`var.ssm_path_prefix` becomes a `chamber` service name.

Reply:
Yup, hopefully flexible enough for us and others.